test/hotspot/jtreg/serviceability/jvmti/ObjectMonitorUsage/ObjectMonitorUsage.java

```diff
@@ -0,0 +1,326 @@
+            check(lockCheck, Thread.currentThread(), 1,
+                  0, // number of threads waiting to enter or re-enter
+                  NUMBER_OF_WAITING_THREADS);
+
+            eThreads = startEnteringThreads(isVirtual);
+
+            // entry count: 1
+            // count of threads waiting to enter:       NUMBER_OF_ENTERING_THREADS
+            // count of threads waiting to re-enter:    0
+            // count of threads waiting to be notified: NUMBER_OF_WAITING_THREADS
+            check(lockCheck, Thread.currentThread(), 1,
+                  NUMBER_OF_ENTERING_THREADS,
+                  NUMBER_OF_WAITING_THREADS);
+
+            for (int i = 0; i < NUMBER_OF_WAITING_THREADS; i++) {
+                lockCheck.notify(); // notify waiting threads one by one
+                // now the notified WaitingTask has to be blocked on the lockCheck re-enter
+
+                // entry count: 1
+                // count of threads waiting to enter:       NUMBER_OF_ENTERING_THREADS
+                // count of threads waiting to re-enter:    i + 1
+                // count of threads waiting to be notified: NUMBER_OF_WAITING_THREADS - i - 1
+                check(lockCheck, Thread.currentThread(), 1,
+                      NUMBER_OF_ENTERING_THREADS + i + 1,
+                      NUMBER_OF_WAITING_THREADS  - i - 1);
+            }
+        }
+        setTestedMonitor(null);
+        joinThreads(wThreads);
+        joinThreads(eThreads);
+        log("###test3: finished " + vtag);
+    }
+
+    static void test(boolean isVirtual) {
+        test0(isVirtual);
+        test1(isVirtual);
+        test2(isVirtual);
+        test3(isVirtual);
+    }
+
+    public static void main(String args[]) {
+        log("\n###main: started\n");
+        check(lockCheck, null, 0, 0, 0);
+
+        test(false); // test platform threads
+        test(true);  // test virtual threads
+
+        check(lockCheck, null, 0, 0, 0);
+        if (getRes() > 0) {
+            throw new RuntimeException("Failed status returned from the agent");
+        }
+        log("\n###main: finished\n");
+    }
+
+    static abstract class TestTask implements Runnable {
+        public volatile boolean ready = false;
+        public abstract void run();
+
+        public void waitReady() {
+            try {
+                while (!ready) {
+                    Thread.sleep(10);
+                }
+            } catch (InterruptedException e) {
+                throw new Error("Unexpected " + e);
+            }
+        }
+    }
+
+    static class EnteringTask extends TestTask {
+        public void run() {
+            ready = true;
+            synchronized (lockCheck) {
+            }
+        }
+    }
+
+    static class WaitingTask extends TestTask {
+         public void run() {
+            synchronized (lockCheck) {
+                try {
+                    ready = true;
+                    lockCheck.wait();
```

Comment:
Spurious wakeups will break this test. They shouldn't happen but ...

Reply:
Thanks, will check how this can be fixed.

Reply:
Sorry I missed this response. I can't see a way to address spurious wakeups in this case as it needs to be a per-thread flag (so that each thread knows it was notified) but you don't know which thread will be notified in any given call to `notify()`. I also can't see how you can detect a spurious wakeup in this code. If they happen then a subtest may fail due to an unexpected number of re-entering threads.
I think we will just have to see how stable the test is in practice.

Reply:
Okay, thanks! I'll add some diagnostic code to catch spurious wakups.
Let's see if we ever encounter any spurious wakeup in this test.

Reply:
I think some sort of comment needs to be added here to document
the possibility of this code path being affect by a spurious wakeup.

Reply:
Thanks. I've added both a comment and also some spurious wakeup detection with a warning.